### PR TITLE
Support for internationalization in Spanish

### PR DESCRIPTION
**Pull request by themonki**
Hi all,

I am from Colombia, I speak Spanish and I have worked in the implementation of OBA for the public transportation system of the city of Cali.  ( [MIO](http://www.mio.com.co) and [Metrocali](http://www.metrocali.gov.co) )

I've made some changes to the front-end pages to support the internationalization using different methods that were already implemented in the tool and i want to share. I hope I can help a little.

Adding values in the URL:

* request_locale
* locale

For example:

`$BASE_URL$/?request_locale=es&locale=es`

The main change in the Core file was in:

* TypedMessagesFactory.java

* GWTMessagesFactory.java

So could read the messages in multiple files "properties".

I have used many keys in properties files, some repeatedly. I do not know if there is a sintaxys established to name the keys to make it easy to change them or find them in the code.

Changes can see them on this page that I have to testing:

[OBA for Cali](http://190.6.160.131:8080/onebusaway-webapp/)
Note: Some pages, such as the principal, not translated, is a static html only as a presentation, be in the other modules. [URL for testing] (http://190.6.160.131:8080/onebusaway-webapp/where/standard/stop.action?id=MIO_512020)

### CODE REVIEW

**Comment by barbeau** [comment]
@themonki Very cool!  I'll have to let others who work more on the web client comment on your PR, but I'm glad to see translations in other languages.

Would you have an interest in submitting a Spanish translation for OneBusAway Android?

It's fairly straightforward, just involves creating two XML files - for adding Spanish, you'd just need to copy the current English arrays/strings XML files:
https://github.com/OneBusAway/onebusaway-android/blob/master/onebusaway-android/src/main/res/values/arrays.xml
https://github.com/OneBusAway/onebusaway-android/blob/master/onebusaway-android/src/main/res/values/strings.xml

...to a new folder `res/values-es/`, and change the English text to Spanish.

As an example, you can see the addition of the initial Finnish translation in two new files in this PR:
https://github.com/OneBusAway/onebusaway-android/pull/253

**Comment by themonki** [comment]
Hi @barbeau 

I will review the links. I would love to collaborate with the Android application.

Do I a PR on [Android] (https://github.com/OneBusAway/onebusaway-android)?

**Comment by barbeau** [comment]
@themonki Yes, for Spanish Android translation you'd submit a PR to the https://github.com/OneBusAway/onebusaway-android project.

Also, for both this project and OBA Android, you'll need to sign the electronic OBA ICLA here before we can merge any changes:
https://docs.google.com/forms/d/12jV-ByyN186MuPotMvxJtNKtSaGGTnEHm8rXomM2bm4/viewform

The same ICLA covers all OBA-related projects (this and OBA Android), so you only need to sign once.

**Comment by themonki** [comment]
@barbeau I have sign the ICLA :+1:

**Comment by barbeau** [comment]
ICLA confirmed :+1: Thanks!

File: onebusaway-webapp/src/main/java/org/onebusaway/webapp/impl/GWTMessagesFactory.java [hunk 1]
```diff
@@ -33,36 +35,34 @@ public GWTMessagesFactory() {
...
-    name = "/" + name.replace('.', '/') + ".properties";
-    InputStream is = _type.getResourceAsStream(name);
-    if (is == null)
-      throw new IllegalStateException("Unable to find resources: " + name);
-    Properties p = new Properties();
-    p.load(is);
-    Handler handler = new Handler(p);
+    name = "/" + name.replace('.', '/');  
+    ReloadableResourceBundleMessageSource messageSource = new ReloadableResourceBundleMessageSource();
+    messageSource.setBasename(name);
```
Comment (themonki) [inline onebusaway-webapp/src/main/java/org/onebusaway/webapp/impl/GWTMessagesFactory.java]: Here we have the base name (eg: ArrivalAndDepartureMessages) and find the locale resources:
* ArrivalAndDepartureMessages_en.properties
* ArrivalAndDepartureMessages_es.properties
* ArrivalAndDepartureMessages_fr.properties
* ArrivalAndDepartureMessages_{lang}.properties

But on this method dont use the default ArrivalAndDepartureMessages.properties file.

File: onebusaway-webapp/src/main/java/org/onebusaway/webapp/impl/TypedMessagesFactory.java [hunk 2]
```diff
@@ -44,43 +46,35 @@ public void setMessagesClass(Class<?> messagesClass) {
...
-    for (Method method : _messagesClass.getDeclaredMethods()) {
-      if (!p.containsKey(method.getName()))
-        _log.warn("missing message name: messagesType="
-            + _messagesClass.getName() + " property=" + method.getName());
-    }
-
-    Handler handler = new Handler(p);
+    name = "/" + name.replace('.', '/');  
+    ReloadableResourceBundleMessageSource messageSource = new ReloadableResourceBundleMessageSource();
+    messageSource.setBasename(name);
```
Comment (themonki) [inline onebusaway-webapp/src/main/java/org/onebusaway/webapp/impl/TypedMessagesFactory.java]: Here we have the base name (eg: ArrivalAndDepartureMessages) and find the locale resources:
* ArrivalAndDepartureMessages_en.properties
* ArrivalAndDepartureMessages_es.properties
* ArrivalAndDepartureMessages_fr.properties
* ArrivalAndDepartureMessages_{lang}.properties

But on this method dont use the default ArrivalAndDepartureMessages.properties file.

File: onebusaway-webapp/src/main/java/org/onebusaway/webapp/gwt/notification/NotificationWidgetPage.java [hunk 3]
```diff
@@ -189,10 +193,11 @@ public Widget update(Context context) throws PageException {
 
     _tripId = context.getParam(PARAM_TRIP_ID);
     _stopId = context.getParam(PARAM_STOP_ID);
+    _serviceDate = context.getParam(PARAM_SERVICE_DATE);
 
-    System.out.println("tripId=" + _tripId + " stopId=" + _stopId);
+    System.out.println("tripId=" + _tripId + " stopId=" + _stopId + " serviceDate=" + _serviceDate);
 
-    if (_tripId == null || _stopId == null) {
+    if (_tripId == null || _stopId == null || _serviceDate == null) {
```
Comment (themonki) [inline onebusaway-webapp/src/main/java/org/onebusaway/webapp/gwt/notification/NotificationWidgetPage.java]: I have a bug with Pop-up on arrival-and-departure-for-stop.action, if this parameter doesn't exist the pop-up don't showing.

Eg on Tampa: Without the parameter serviceDate: 

`http://tampa.onebusaway.org/where/standard/arrival-and-departure-for-stop.action?stopId=Hillsborough+Area+Regional+Transit_3165&tripId=Hillsborough+Area+Regional+Transit_119590&vehicleId=Hillsborough+Area+Regional+Transit_2217&stopSequence=70`

With the parameter serviceDate:

`http://tampa.onebusaway.org/where/standard/arrival-and-departure-for-stop.action?stopId=Hillsborough+Area+Regional+Transit_3165&tripId=Hillsborough+Area+Regional+Transit_119590&serviceDate=1452142800000&vehicleId=Hillsborough+Area+Regional+Transit_2217&stopSequence=70`

Then I added this validation.

File: onebusaway-webapp/src/main/resources/org/onebusaway/webapp/gwt/AgenciesMap.gwt.xml [hunk 4]
```diff
@@ -43,5 +43,9 @@
 
 	<!-- Specify the app entry point class -->
 	<entry-point class='org.onebusaway.webapp.gwt.agencies_map.AgenciesMap' />
+	
+	<!-- Availabe Locales -->
+	<extend-property name="locale" values="en"/>
+	<extend-property name="locale" values="es"/>
```
Comment (themonki) [inline onebusaway-webapp/src/main/resources/org/onebusaway/webapp/gwt/AgenciesMap.gwt.xml]: For GWT components, it is necessary to add all supported languages on some gwt.xml.

Not sure if you only need to add one or all files.

**Comment by themonki** [comment]
Hi,

I do not know if someone has reviewed this. It would be nice label with enhancement.

I recreated (deleted and created again) fork to create a new branch (although it may have created it without doing that, my ignorance):

https://github.com/themonki/onebusaway-application-modules/tree/internationalization

I do not know whether to create a new PR or better wait for any comments? I've made some changes to the format according to the [style guide](https://github.com/OneBusAway/onebusaway/wiki/Code-Style), but still have the[ same commits](https://github.com/OneBusAway/onebusaway-application-modules/compare/master...themonki:internationalization?expand=1).

**Comment by barbeau** [comment]
@themonki If you have new changes, and have previously deleted your Github fork, I believe you'll need to create a new pull request (and close this one).

**Comment by barbeau** [comment]
Closed in favor of https://github.com/OneBusAway/onebusaway-application-modules/pull/164

**Pull request closed without merging**